### ISSUE TITLE
Suggestion: add Coverity Scan... and fix most important issues

### ISSUE THREAD

**Issue by GitMensch** [LADSoft/OrangeC#63]
Note: This likely is an issue that needs some time, especially inspecting and fixing the results.
https://scan.coverity.com

But I'd recommend it as it helps to fix CRT or other errors before they actually happen...

**Comment by LADSoft** [comment]
making progress, will see if I can make a yml file for TRAVIS CI tomorrow

**Comment by GitMensch** [comment]
It is nice to have https://travis-ci.org/LADSoft/OrangeC working in the future. Will using GCC directly work or only with the [mingw gcc](https://lists.boost.org/Archives/boost/2013/06/204433.php) version?

Will [compilation on osx](https://docs.travis-ci.com/user/reference/osx) work, too?

Did you tried a local run of the coverity scan using cl.exe and manual upload of the resulting file already?

**Comment by LADSoft** [comment]
I'm not sure yet on the travis-ci...    I didn't realize what I was getting into :).    The main problem with compiling for other operating systems is rewriting the runtime library...   I've just never gotten there...

I tried a local run of coverity...  the coverage is only at 72% so it wouldn't be accepted.   A couple of the other issues that have been opened are already long-term issues so I'll get to it as soon as I can...   just leave this open to keep me honest :)

**Comment by GitMensch** [comment]
> the coverage is only at 72% so it wouldn't be accepted

I *guess* it works if you at least temporarily) remove sources that aren't compiled, wouldn't it?

**Comment by LADSoft** [comment]
I went and looked at the problem, the COV program is having a lot of trouble with compiling references to std::tuple in various places throughout the project.   That is going to be a major issue so I'm going to defer for a bit...   will get to it soon :)

**Comment by GitMensch** [comment]
... What do you mean with "the COV program is having a lot of trouble compiling"?

You just call `cov-build` (maybe with an additional `cov-configure` for adjusting to use cl.exe instead of gcc) and everything is fine, isn't it?

see lower parts of https://scan.coverity.com/download?tab=cxx

**Comment by LADSoft** [comment]
cov-translate throws errors on sources that are compilable by all the compilers.   IT does it in a lot of places.   For example:

Errors.cpp 
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|> "C:\Program Files (x86)\Microsoft Visual
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>           Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\tuple", line
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>           864: error #494: no instance of overloaded function
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>           "std::tuple<_This, _Rest...>::tuple" matches the specified type
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>   		constexpr tuple<_This, _Rest...>::tuple(_Tag, _Tpl&& _Right, index_sequence<_Indices...>)
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>   		                                  ^
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|> 
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|> "C:\Program Files (x86)\Microsoft Visual
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>           Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\tuple", line
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>           876: error #494: no instance of overloaded function
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>           "std::tuple<_This, _Rest...>::tuple" matches the specified type
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>   		tuple<_This, _Rest...>::tuple(_Tag, const _Alloc& _Al, _Tpl&& _Right, index_sequence<_Indices...>)
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|>   		                        ^
2018-05-04T21:43:05.737863Z|cov-translate|13616|output|> 
2018-05-04T21:43:06.829266Z|cov-translate|13616|output|> Emit for file 'C:/orangec/src/ocpp/Errors.cpp' complete.
2018-05-04T21:43:06.880712Z|cov-translate|13616|output|> [ERROR] 2 errors detected in the compilation of "Errors.cpp".

**Comment by GitMensch** [comment]
Interesting, seems like it is definitely good to talk with the coverity scan people about this issue.

Is `cov-translate` internally called by `cov-build`?

**Comment by LADSoft** [comment]
as far as I know...   all I did was run cov-build.   Although the configuration was kind of dicey, I wasn't 100% sure what was going on there but I thought I had it set up properly.   Seemed to find the MSVC include directories though...

**Comment by LADSoft** [comment]
The latest on this is that Monday someone responded and said they were forwarding to another group; that they would get back to me soon.   I'm still waiting...

**Comment by LADSoft** [comment]
still no answer.  I asked again.   If nothing moves on this in another week or so I will look into trying to configure my setup to use mingw-w64 instead of the microsoft compiler.

**Comment by LADSoft** [comment]
it compiled successfully with mingw64 and is now submitted.

**Comment by LADSoft** [comment]
the submission came up with 736 defects with a defect density of 1.40.   The worst of them were potential buffer overruns(120), which really doesn't surprise me as that is one of the areas I'm weak in.   I haven't figured out yet how to see the actual defects but in the meantime did configure it so I can get an idea of which subprojects are most problematic...

**Comment by LADSoft** [comment]
90% of the defects are in occ and ocide, which is under half the code base...   there are relatively few problems in the C++ code which is pretty much all the other projects...   meanwhile coverity says that have to check me out and make sure i'm for real before they will let me view the actual defects :)

**Comment by GitMensch** [comment]
@LADSoft wrote in https://github.com/LADSoft/OrangeC/pull/232#issuecomment-401214518

>> @LADSoft Sounds like a good time to upload a new version to Coverity Scan [...]

> @GitMensch will do as soon as I get the codebase reformatting done (should be today or tomorrow)

Do you have any news about this?
What do you think about adding a Badge to the README? The number may look high but it is always better to stay honest, other projects are worse, and the most important part for now on this issue is actually to just not increase the number + fix the CRT issues (as I understood you there are not many in there and the CRT is the part that has the biggest impact once valid source got compiled).

**Comment by LADSoft** [comment]
I took a bit of a holiday.   One good part about doing open-source, don't have to stick to high pressure schedules when i don't want to :smile:

Tried adding a badge to the README for the latest scan but failed, anything you can point out?   If not I'll look at it again tomorrow.

The latest scan found 23 new issues but fixed 26.   Go figure!

I guess to invite anyone to take a look I need to have an email address for the invite?   I couldn't find a way to invite based on github user anyway...   so unless you have another suggestion anyone who wants an invite should get me an email address...

**Comment by GitMensch** [comment]
The README is in. There currently is no CI based recheck - I think going back to [Travis](https://travis-ci.org/LADSoft/OrangeC) and try to add a a hook to auto-build (and possibly automatically submit to coverity scan) would be good.

Do you remember if/where issues with Travis exists?

@chuggafan: You may consider to [join this project@coverity](https://scan.coverity.com/memberships/new?project_id=ladsoft-orangec)

**Comment by LADSoft** [comment]
I don't think there were any specific issues with travis, just at the time I was doing it the coverity thing was all very new and I wasn't sure which way to proceed...  I have a little more experience now so I can give this a try.

Order for the things on my list are:   finish #242, then deal with the two pull requests from @chuggafan, then do this and the issue you brought up with adding a build via assembly to the appveyor tests, then fix any issues beyond #242 that chuggafan needs to proceed, then come back to this issue and start working on the most serious problems.   After that I guess I will work with @chuggafan on anything else he needs to proceed and then get back to fixing the floating point math problems.

**Comment by LADSoft** [comment]
the current progress on travis-ci is that after a bunch of trial and error I finally got the webhook readded, but github and travis CI are both in a state of flux right now because github couldn't deliver webhooks for a while and there is a lot of backlog.   So I'll get back to this and see if I can finish it out in a couple of days when they have things fixed :)   Just as a point of interest coverity has scan limits of something like one a day so we can't scan every push, but I think just having it available through travis-ci will make things a lot more streamlined .

**Comment by LADSoft** [comment]
I'm a little confused because the coverity scan completed on travis-ci but the upload message seemed to say the upload both failed and completed!    But I looked on coverity site and something is queued so I guess it worked :)

**Comment by GitMensch** [comment]
> But I looked on coverity site and something is queued so I guess it worked :)

Yes, it worked, if I remember correctly > 200 issues fixed, > 100 new issues. So in overal: got better.
Just to recheck: the first upload on coverity was also without OCIDE checked, correct?

**Comment by LADSoft** [comment]
the original uploads had OCIDE checked...

**Comment by GitMensch** [comment]
So likely no actual drop in the number of issues, only issues "not checked any more" :-(
But OCIDE *really* should be left out until the rest is considered to be fixed.

**Comment by LADSoft** [comment]
I am planning on chipping away at this issue for a while this weekend.

**Comment by LADSoft** [comment]
well even though coverity was down I made some progress - CLANG warns about some of the more obvious things coverity would catch so I went through the CLANG and MINGW64 warning logs and cleaned them up.   

   Still to do in that vein is work on OCIDE - CLANG caught some outright bugs which I was hesitant to fix on the quick runthrough I was doing, because I don't yet know the impact that will have on the behavior of the program.

**Comment by LADSoft** [comment]
I think I read that coverity is up again; couldn't find the readonly banners when i looked.   As soon as I get the current project (unique_ptr) done I will try a scan to see where we stand.

**Comment by GitMensch** [comment]
>  As soon as I get the current project (unique_ptr) done I will try a scan to see where we stand.

Ping

**Comment by LADSoft** [comment]
I'm gonna break this out into several related issues to make it more manageable.   Doing it all at once isn't going to work but tracking on a per-milestone basis would be good.

**Comment by GitMensch** [comment]
This is very reasonable and it is good to have the libs included in the current milestone.
Could be closed as coverity-scan was added and we have issues for the different parts now with #392 #393 #394 #395 #397 and least important #396.

One thing that *may* be done before closing this one: update the coverity results (the last update was from Oct 30, 2018).

**Comment by LADSoft** [comment]
I'm working on getting coverity up and running again.   I need to add the new occil related projects to the scan and to do that I have to run a program on my PC...   it is up in the air right now though as the AV program I am using decided to send the program in to the AV company lab for testing.  It triggered some rare behavior they apparently check for.

**Comment by GitMensch** [comment]
I've assumed the msil components to be the reason for the new defects, but couldn't find it as a component. Which component is new (or are those parts currently in the "others")?

**Comment by GitMensch** [comment]
> I'm trying to get the MSIL library as a separate project in coverity. Coverity says it may take a while to update, though...

Yes, these filters are sometimes "special" (the definition page says "Component changes will be reflected automatically within 15 minutes after submission. There is no need to submit a new build to reflect the changes.", those 15 minutes passed and I didn't see anything msil when checking). It might be possible to include the msil backend in the project by specifying the related files in the msil project. (you may want to contact the support to recheck if there's a need to order these components, or just try it out [in theory the lines-of-code should decrease in occ and increase in msil]).
I think it is very reasonable to use an own component for sqlite3 and ignore the component because it is an external part [which should be updated from 3.7.12.1 to the current 3.29.0, see [Release History of SQLite](https://sqlite.org/changes.html), but this likely should get a separate issue)
